File: go/internal/bench/handoff_test.go
package bench

import (
	"testing"
	"time"
)

func TestMeasureHandoffSampleCountAndCalls(t *testing.T) {
	cfg := HandoffConfig{Warmup: 3, Iterations: 5, RingCap: 16}
	calls := 0
	samples := MeasureHandoff(cfg, func() { calls++ })
	if len(samples) != 5 {
		t.Fatalf("want 5 samples, got %d", len(samples))
	}
	if calls != 8 {
		t.Fatalf("want 8 calls (warmup+iterations), got %d", calls)
	}
}

func TestMeasureHandoffPacedKeepsTheGapBetweenRoundTrips(t *testing.T) {
	gap := int64(time.Millisecond)
	cfg := HandoffConfig{Warmup: 1, Iterations: 4, RingCap: 16, GapNs: gap}
	var starts []time.Time
	samples := MeasureHandoffPaced(cfg, func() { starts = append(starts, time.Now()) })
	if len(samples) != 4 {
		t.Fatalf("want 4 samples, got %d", len(samples))
	}
	if len(starts) != 5 {
		t.Fatalf("want 5 calls (warmup+iterations), got %d", len(starts))
	}
	for i := 1; i < len(starts); i++ {
		if d := starts[i].Sub(starts[i-1]).Nanoseconds(); d < gap {
			t.Fatalf("round trips %d..%d only %dns apart, want >= %dns", i-1, i, d, gap)
		}
	}
	// The gap busy-wait must sit OUTSIDE the timed window: each sample is the
	// (near-instant) round trip, never the millisecond gap.
	for i, s := range samples {
		if s >= gap {
			t.Fatalf("sample %d = %dns includes the gap", i, s)
		}
	}
}

File: java/common/src/test/java/net/knego/hiperf/common/HandoffTest.java
package net.knego.hiperf.common;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

class HandoffTest {

    @Test
    void measureRunsWarmupPlusIterationsAndReturnsIterationsSamples() {
        HandoffConfig cfg = new HandoffConfig(3, 5, 16, 0);
        AtomicInteger calls = new AtomicInteger();
        long[] samples = Handoff.measure(cfg, calls::incrementAndGet);
        assertEquals(5, samples.length, "one sample per measured iteration");
        assertEquals(8, calls.get(), "warmup (3) + iterations (5) calls");
    }

    @Test
    void measurePacedKeepsTheGapBetweenRoundTripsAndOutsideSamples() {
        long gapNs = 1_000_000; // 1 ms: far above clock granularity
        HandoffConfig cfg = new HandoffConfig(1, 4, 16, gapNs);
        List<Long> starts = new ArrayList<>();
        long[] samples = Handoff.measurePaced(cfg, () -> starts.add(System.nanoTime()));
        assertEquals(4, samples.length);
        assertEquals(5, starts.size(), "warmup (1) + iterations (4) calls");
        for (int i = 1; i < starts.size(); i++) {
            long d = starts.get(i) - starts.get(i - 1);
            assertTrue(d >= gapNs, "round trips only " + d + "ns apart, want >= " + gapNs);
        }
        // The gap busy-wait sits OUTSIDE the timed window: each sample is the
        // near-instant round trip, never the millisecond gap.
        for (long s : samples) {
            assertTrue(s < gapNs, "sample " + s + "ns includes the gap");
        }
    }
}
